In [1]:
# ===== 기본 세트 (거의 항상 필요) =====
import pandas as pd                    # 데이터프레임 다루기 (표 형태 데이터 처리)
import numpy as np                     # 수치 계산, 배열 연산
import matplotlib.pyplot as plt        # 기본 시각화
import matplotlib.font_manager as fm   # 그래프에 한글 폰트 적용 시 필요
import seaborn as sns                  # matplotlib 기반의, 더 예쁘고 간결한 통계 시각화

# ===== 한글 폰트 설정 =====
font_path = "C:/Windows/Fonts/malgun.ttf"
fm.fontManager.addfont(font_path)
plt.rcParams["font.family"] = fm.FontProperties(fname=font_path).get_name()
plt.rcParams["axes.unicode_minus"] = False

# ===== 파일/경로/시스템 =====
import os                              # 파일 경로 확인, 파일 존재 여부(os.path.exists) 등
from pathlib import Path               # 폴더/파일 경로를 객체로 다루기 (work_dir / "파일명")
import joblib                          # 학습된 scaler/encoder를 파일로 저장하고 재사용
import platform                        # OS 종류 확인 (Windows/Mac/Linux 구분 필요할 때)
import warnings                        # 경고 메시지 제어(숨기기/표시)
import matplotlib.pyplot as plt
import urllib.request, zipfile

# ===== 벤치마크/성능 측정 =====
import time                            # 코드 실행 시간 측정 (time.perf_counter())
import psutil                          # 메모리 사용량 측정 (rss_mb 등)
import gc                              # 메모리 강제 정리(garbage collection)

# ===== 결측치 탐색 시각화 =====
import missingno as msno               # 결측치 위치(matrix), 관계(heatmap) 시각화

# ===== 인터랙티브 시각화 =====
import plotly.express as px            # 마우스 hover/zoom 가능한 인터랙티브 그래프
import plotly.io as pio                # plotly 렌더러(출력 방식) 설정

# ===== 대용량 데이터 처리 =====
import polars as pl                    # pandas보다 빠른 대용량 데이터 처리 (Rust 기반)

# ===== 통계 검증 =====
import statsmodels.api as sm           # 회귀분석, 다중공선성(Cond. No.) 등 통계 검증
import statsmodels.formula.api as smf  # R 스타일 수식("y ~ x1 + x2")으로 회귀분석
from scipy import stats                # t-검정·카이제곱·정규성 검정 등 통계 검정 모음
from statsmodels.stats.power import TTestIndPower   # 검정력 분석(필요 표본수, 검정력 계산)
from sklearn.neighbors import NearestNeighbors      # 최근접 이웃 탐색 (경향점수 매칭 등에 사용)

# ===== 전처리(인코딩/스케일링) =====
from sklearn.preprocessing import LabelEncoder      # 카테고리를 숫자로 변환 (순서형)
from sklearn.preprocessing import MinMaxScaler      # 0~1 범위로 스케일링
from sklearn.preprocessing import StandardScaler    # 평균0, 표준편차1로 스케일링
from sklearn.preprocessing import RobustScaler      # 이상치에 강한 스케일링 (median/IQR 기반)

# ===== 모델 =====
from sklearn.linear_model import LinearRegression       # 선형 회귀 모델
from sklearn.linear_model import LogisticRegression     # 로지스틱 회귀(분류) 모델
from sklearn.tree import DecisionTreeClassifier         # 의사결정나무 분류 모델
from sklearn.dummy import DummyClassifier               # 베이스라인(단순 규칙) 분류 모델
from sklearn.ensemble import RandomForestRegressor      # 랜덤포레스트 회귀 모델
from sklearn.ensemble import RandomForestClassifier     # 랜덤포레스트 분류 모델(여러 트리의 앙상블)
from xgboost import XGBClassifier                       # XGBoost 분류 모델(부스팅 기반)

# ===== 파이프라인 =====
from sklearn.pipeline import make_pipeline              # 전처리+모델 여러 단계를 하나로 묶기

# ===== 데이터 분리 / 교차검증 =====
from sklearn.model_selection import train_test_split    # 데이터를 학습용/테스트용으로 분리
from sklearn.model_selection import KFold               # 데이터를 K개로 분할(교차검증용, 회귀에 적합)
from sklearn.model_selection import StratifiedKFold     # 클래스 비율을 유지하며 K개로 분할(불균형 분류용)
from sklearn.model_selection import cross_val_score     # 교차검증으로 단일 지표 반복 평가
from sklearn.model_selection import cross_validate      # 교차검증으로 여러 지표·학습점수까지 한번에 평가
from sklearn.model_selection import learning_curve      # 표본 크기에 따른 학습/검증 성능 변화 확인

# ===== 평가지표 =====
from sklearn.metrics import accuracy_score              # 분류 모델의 정확도 계산
from sklearn.metrics import confusion_matrix            # 실제/예측 결과를 교차표로 정리(TP/FP/TN/FN)
from sklearn.metrics import classification_report       # 정밀도·재현율·F1 등을 한번에 요약
from sklearn.metrics import precision_score             # 정밀도(예측한 것 중 실제로 맞은 비율)
from sklearn.metrics import recall_score                # 재현율(실제 정답 중 모델이 잡아낸 비율)
from sklearn.metrics import f1_score                     # 정밀도와 재현율의 조화평균
from sklearn.metrics import roc_curve                    # ROC 곡선 좌표 계산
from sklearn.metrics import roc_auc_score                # ROC 곡선 아래 면적(AUC)
from sklearn.metrics import precision_recall_curve       # 정밀도-재현율 곡선 좌표 계산
from sklearn.metrics import average_precision_score      # PR 곡선 아래 면적(AP)
from sklearn.metrics import log_loss                      # 분류 확률 예측의 손실(로그 손실)


from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import OrdinalEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import TargetEncoder
from sklearn.model_selection import GridSearchCV

In [2]:
# ─────────────────────────────────────────────
# [C1] ⚙️ 데이터 준비
# 최초 1회 다운로드 → data/ 폴더에 저장 (이후 오프라인)
# ─────────────────────────────────────────────
import urllib.request, zipfile
from pathlib import Path

import numpy as np
import pandas as pd

RANDOM_STATE = 42
np.random.seed(RANDOM_STATE)

DATA_DIR = Path("data")
DATA_DIR.mkdir(exist_ok=True)

def fetch_uci(url, zip_name, member):
    """UCI 정적 저장소의 zip을 내려받아 data/에 풀고, CSV 경로를 돌려줍니다."""
    csv_path = DATA_DIR / member
    if not csv_path.exists():
        zip_path = DATA_DIR / zip_name
        if not zip_path.exists():
            print(f"내려받는 중… {zip_name}")
            urllib.request.urlretrieve(url, zip_path)
        with zipfile.ZipFile(zip_path) as z:
            z.extract(member, DATA_DIR)
    return csv_path

shoppers = pd.read_csv(fetch_uci(
    "https://archive.ics.uci.edu/static/public/468/online+shoppers+purchasing+intention+dataset.zip",
    "online_shoppers.zip", "online_shoppers_intention.csv"))

NUM_COLS = ["Administrative", "Administrative_Duration", "Informational",
            "Informational_Duration", "ProductRelated", "ProductRelated_Duration",
            "BounceRates", "ExitRates", "PageValues", "SpecialDay"]
CAT_COLS = ["Month", "OperatingSystems", "Browser", "Region",
            "TrafficType", "VisitorType", "Weekend"]

y = shoppers["Revenue"].astype(int)

print(f"수치형 {len(NUM_COLS)}개 · 범주형 {len(CAT_COLS)}개 · 타깃 1개")
print("범주별 값 개수:", {c: shoppers[c].nunique() for c in CAT_COLS})
print("\n→ 준비 완료. 이제 여러분 차례입니다.")

내려받는 중… online_shoppers.zip
수치형 10개 · 범주형 7개 · 타깃 1개
범주별 값 개수: {'Month': 10, 'OperatingSystems': 8, 'Browser': 13, 'Region': 9, 'TrafficType': 20, 'VisitorType': 3, 'Weekend': 2}

→ 준비 완료. 이제 여러분 차례입니다.


In [10]:
#[문제 1]
#1) ColumnTransformer를 만듭니다.
#   - 수치형: "passthrough"
#   - 범주형: OneHotEncoder(handle_unknown="ignore", min_frequency=20)
#2) 그것과 지난 순서의 모델을 Pipeline으로 묶습니다.
#   RandomForestClassifier(n_estimators=300, min_samples_leaf=20)
#3) 5겹 CV로 F1과 AP(`average_precision`)를 측정하고, 수치형 10개만 썼을 때와 비교합니다.
#4) 변환 후 피처가 몇 개가 됐는지 출력합니다.

# [C2] 문제 1. 범주형 7개 열을 `Pipeline`으로 푼다
# ⌨️ 문제 1 — ColumnTransformer + Pipeline으로 범주형 개방
# 여기에 코드를 작성하세요 (ColumnTransformer → Pipeline → CV 비교 → 피처 수)
X = shoppers[NUM_COLS + CAT_COLS]
y = shoppers["Revenue"].astype(int)

X_train, X_test, y_train, y_test =train_test_split(
    X, y, test_size = 0.2, stratify= y, random_state = 42)

preprocess = ColumnTransformer([
    ("num", "passthrough", NUM_COLS),
    ("cat", OneHotEncoder(handle_unknown="ignore", min_frequency=20), CAT_COLS)
])

pipe = Pipeline([
    ("pre", preprocess),
    ("clf", RandomForestClassifier(n_estimators=300, min_samples_leaf=20, random_state=42))
])

skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
scoring = ["f1", "average_precision"]

results = cross_validate(pipe, X_train, y_train, cv = skf, scoring=scoring)

for metric in scoring :
    print(f"{metric}: {results[f'test_{metric}'].mean():.4f}")

preprocess.fit(X_train)
n_features_after = preprocess.transform(X_train).shape[1]
print(f"\n변환 후 피처 개수: {n_features_after}개")

f1: 0.5714
average_precision: 0.7381

변환 후 피처 개수: 66개


In [20]:
#[문제 2]
#1) 아래 두 가설을 코드로 만들기 전에, 각각 "왜 효과가 있을 것인가"를 한 줄로 적습니다.
#   가설 A: has_page_value = (PageValues > 0)  ← 0 여부가 추가 표현으로 유용할 수 있다
#   가설 B: total_pages = Administrative + Informational + ProductRelated  ← 총 탐색량
#2) 문제 1의 ② 구성을 기준으로, A만 / B만 각각 추가해 5겹 CV로 측정합니다.
#3) 각각의 ΔF1과 ΔAP를 계산해 채택·기각을 결정합니다.

# [C3] 문제 2. 파생변수 — 가설을 먼저 적고, 그다음 검증한다
# ⌨️ 문제 2 — 가설 A·B를 각각 따로 검증

# 여기에 코드를 작성하세요 (파생 열 추가 → A만 / B만 → Δ 계산 → 채택 결정)

shoppers["has_page_value"] = (shoppers["PageValues"] > 0).astype(int)
shoppers["total_pages"] = shoppers["Administrative"] + shoppers["Informational"] + shoppers["ProductRelated"]

X = shoppers[NUM_COLS + CAT_COLS + ["has_page_value", "total_pages"]]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42)

def cv_of(num_cols, cat_cols):
    pre = ColumnTransformer([
        ("num", "passthrough", num_cols),
        ("cat", OneHotEncoder(handle_unknown="ignore", min_frequency=20), cat_cols)
    ])
    pipe = Pipeline([
        ("pre", pre),
        ("clf", RandomForestClassifier(n_estimators=300, min_samples_leaf=20, random_state=42))
    ])
    results = cross_validate(pipe, X_train[num_cols + cat_cols], y_train, cv=skf, scoring=scoring)
    return {m: results[f"test_{m}"].mean() for m in scoring}


base = cv_of(NUM_COLS, CAT_COLS)
print(f"② 기준: F1={base['f1']:.4f}, AP={base['average_precision']:.4f}")

# 가설 A만 추가
score_a = cv_of(NUM_COLS + ["has_page_value"], CAT_COLS)
delta_a_f1 = score_a["f1"] - base["f1"]
delta_a_ap = score_a["average_precision"] - base["average_precision"]
print(f"가설 A 추가: F1={score_a['f1']:.4f}(Δ{delta_a_f1:+.4f}), AP={score_a['average_precision']:.4f}(Δ{delta_a_ap:+.4f})")

# 가설 B만 추가
score_b = cv_of(NUM_COLS + ["total_pages"], CAT_COLS)
delta_b_f1 = score_b["f1"] - base["f1"]
delta_b_ap = score_b["average_precision"] - base["average_precision"]
print(f"가설 B 추가: F1={score_b['f1']:.4f}(Δ{delta_b_f1:+.4f}), AP={score_b['average_precision']:.4f}(Δ{delta_b_ap:+.4f})")

# 3) 채택/기각 결정
print("\n=== 채택/기각 결정 ===")
for name, d_f1, d_ap in [("가설 A(has_page_value)", delta_a_f1, delta_a_ap),
                          ("가설 B(total_pages)", delta_b_f1, delta_b_ap)]:
    decision = "채택" if (d_f1 > 0 and d_ap > 0) else "기각"
    print(f"{name}: {decision} (ΔF1={d_f1:+.4f}, ΔAP={d_ap:+.4f})")

② 기준: F1=0.5714, AP=0.7381
가설 A 추가: F1=0.6458(Δ+0.0744), AP=0.7533(Δ+0.0153)
가설 B 추가: F1=0.5667(Δ-0.0047), AP=0.7398(Δ+0.0017)

=== 채택/기각 결정 ===
가설 A(has_page_value): 채택 (ΔF1=+0.0744, ΔAP=+0.0153)
가설 B(total_pages): 기각 (ΔF1=-0.0047, ΔAP=+0.0017)


In [21]:
# [C4] 문제 3. 누수를 직접 만들어, 부풀림을 잰다
# ⌨️ 문제 3 — 같은 피처를 누수 버전과 정상 버전으로 각각 재기
from sklearn.preprocessing import TargetEncoder

# 여기에 코드를 작성하세요 (조합 키 → 누수 버전 CV → Pipeline 안 버전 CV → 차이)

#[문제 3]
#1) 조합 키를 만듭니다 — combo = Month + "_" + TrafficType + "_" + Region
#   (몇 개 조합이 나오는지, 10건 미만인 조합이 몇 개인지 확인합니다)
#2) 누수 버전: 전체 데이터로 조합별 Revenue 평균을 구해 열로 추가 → CV
#3) 교차 적합 버전: 같은 일을 `TargetEncoder`로 `Pipeline` 안에서 수행 → CV
#4) 두 점수의 차이를 계산하고, 교차 적합 버전이 기준(문제 1의 ②)보다 나아졌는지 확인합니다.

# 1) 조합 키 만들기
shoppers["combo"] = (shoppers["Month"].astype(str) + "_" +
                     shoppers["TrafficType"].astype(str) + "_" +
                     shoppers["Region"].astype(str))

n_combos = shoppers["combo"].nunique()
combo_counts = shoppers["combo"].value_counts()
n_small = (combo_counts < 10).sum()
print(f"전체 조합 개수: {n_combos}개")
print(f"10건 미만인 조합 개수: {n_small}개")

X_full = shoppers[NUM_COLS + CAT_COLS + ["combo"]]
X_train, X_test, y_train, y_test = train_test_split(X_full, y, test_size=0.2, stratify=y, random_state=42)

skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# 2) 누수 버전: 전체 데이터로 조합별 Revenue 평균을 미리 계산해서 열로 추가
combo_target_mean_leak = shoppers.groupby("combo")["Revenue"].mean()   # 전체(train+test)로 계산 → 누수!
X_train_leak = X_train.copy()
X_train_leak["combo_encoded_leak"] = X_train_leak["combo"].map(combo_target_mean_leak)

pre_leak = ColumnTransformer([
    ("num", "passthrough", NUM_COLS + ["combo_encoded_leak"]),
    ("cat", OneHotEncoder(handle_unknown="ignore", min_frequency=20), CAT_COLS)
])
pipe_leak = Pipeline([
    ("pre", pre_leak),
    ("clf", RandomForestClassifier(n_estimators=300, min_samples_leaf=20, random_state=42))
])
score_leak = cross_val_score(
    pipe_leak, X_train_leak[NUM_COLS + CAT_COLS + ["combo_encoded_leak"]], y_train, cv=skf, scoring="f1"
).mean()
print(f"누수 버전 F1: {score_leak:.4f}")

# 3) 교차 적합 버전: TargetEncoder를 Pipeline 안에 포함 (자동으로 누수 방지 처리됨)
pre_correct = ColumnTransformer([
    ("num", "passthrough", NUM_COLS),
    ("cat", OneHotEncoder(handle_unknown="ignore", min_frequency=20), CAT_COLS),
    ("combo", TargetEncoder(random_state=42), ["combo"])
])
pipe_correct = Pipeline([
    ("pre", pre_correct),
    ("clf", RandomForestClassifier(n_estimators=300, min_samples_leaf=20, random_state=42))
])
score_correct = cross_val_score(
    pipe_correct, X_train[NUM_COLS + CAT_COLS + ["combo"]], y_train, cv=skf, scoring="f1"
).mean()
print(f"교차 적합(TargetEncoder in Pipeline) 버전 F1: {score_correct:.4f}")
print(f"누수로 인한 부풀림: {score_leak - score_correct:+.4f}")

# 4) 기준(② NUM+CAT)과 비교
pre_base = ColumnTransformer([
    ("num", "passthrough", NUM_COLS),
    ("cat", OneHotEncoder(handle_unknown="ignore", min_frequency=20), CAT_COLS)
])
pipe_base = Pipeline([
    ("pre", pre_base),
    ("clf", RandomForestClassifier(n_estimators=300, min_samples_leaf=20, random_state=42))
])
score_base = cross_val_score(pipe_base, X_train[NUM_COLS + CAT_COLS], y_train, cv=skf, scoring="f1").mean()

print(f"기준(② NUM+CAT) F1: {score_base:.4f}")
print(f"정상 버전(combo 추가) F1: {score_correct:.4f}")
print(f"개선 여부: {'개선됨' if score_correct > score_base else '개선 안됨'} ({score_correct - score_base:+.4f})")

전체 조합 개수: 715개
10건 미만인 조합 개수: 464개
누수 버전 F1: 0.5963


c:\Users\pyostry\.conda\envs\myenv\Lib\site-packages\sklearn\preprocessing\_target_encoder.py:341: FutureWarning: `TargetEncoder.shuffle` and `TargetEncoder.random_state` are deprecated in version 1.9 and will be removed in version 1.11. Pass a cross-validation generator as `cv` argument to specify the shuffling behaviour instead.
  warnings.warn(
c:\Users\pyostry\.conda\envs\myenv\Lib\site-packages\sklearn\preprocessing\_target_encoder.py:341: FutureWarning: `TargetEncoder.shuffle` and `TargetEncoder.random_state` are deprecated in version 1.9 and will be removed in version 1.11. Pass a cross-validation generator as `cv` argument to specify the shuffling behaviour instead.
  warnings.warn(
c:\Users\pyostry\.conda\envs\myenv\Lib\site-packages\sklearn\preprocessing\_target_encoder.py:341: FutureWarning: `TargetEncoder.shuffle` and `TargetEncoder.random_state` are deprecated in version 1.9 and will be removed in version 1.11. Pass a cross-validation generator as `cv` argument to specify 

교차 적합(TargetEncoder in Pipeline) 버전 F1: 0.5589
누수로 인한 부풀림: +0.0373
기준(② NUM+CAT) F1: 0.5714
정상 버전(combo 추가) F1: 0.5589
개선 여부: 개선 안됨 (-0.0125)


In [22]:
# [C5] 문제 4. `Pipeline` 통째로 `GridSearchCV`
# ⌨️ 문제 4 — 전처리 옵션과 모델 옵션을 한 격자에서
# 여기에 코드를 작성하세요 (격자 정의 → GridSearchCV → 결과표 → best)

#[문제 4]
#1) 문제 2에서 채택한 구성(수치 10개 + has_page_value + 범주 7개)으로 Pipeline을 만듭니다.
#2) GridSearchCV로 다음 두 손잡이를 함께 탐색합니다. `scoring="average_precision"`
#   - pre__cat__min_frequency: [10, 50]      ← 전처리 옵션
#   - clf__min_samples_leaf:   [5, 20, 50]   ← 모델 옵션
#3) 조합별 내부 CV 점수를 표로 출력하고, `best_params_`와 `best_score_`를 확인합니다.
#4) 선택된 모델의 F1도 같은 5겹 분할에서 측정해 기록합니다. 이 값은 모델 선택에 사용한 데이터의 내부 추정치이므로 최종 일반화 성능으로 보고하지 않습니다.

# 0) 문제 2에서 채택한 구성으로 데이터 준비
FEAT_NUM = NUM_COLS + ["has_page_value"]

X_full = shoppers[FEAT_NUM + CAT_COLS]
y = shoppers["Revenue"].astype(int)

X_train, X_test, y_train, y_test = train_test_split(
    X_full, y, test_size=0.2, stratify=y, random_state=42
)

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# 1) Pipeline 만들기
pre = ColumnTransformer([
    ("num", "passthrough", FEAT_NUM),
    ("cat", OneHotEncoder(handle_unknown="ignore"), CAT_COLS)   # min_frequency는 grid에서 탐색할 것이라 여기선 기본만
])
pipe = Pipeline([
    ("pre", pre),
    ("clf", RandomForestClassifier(n_estimators=300, random_state=42))
])

# 2) GridSearchCV — 전처리 옵션 + 모델 옵션 동시 탐색
param_grid = {
    "pre__cat__min_frequency": [10, 50],
    "clf__min_samples_leaf": [5, 20, 50]
}

gs = GridSearchCV(pipe, param_grid, cv=cv, scoring="average_precision", n_jobs=-1)
gs.fit(X_train, y_train)

# 3) 결과표 출력
results_df = pd.DataFrame(gs.cv_results_)[
    ["param_pre__cat__min_frequency", "param_clf__min_samples_leaf", "mean_test_score", "std_test_score"]
]
results_df = results_df.sort_values("mean_test_score", ascending=False)
print("=== 조합별 CV 점수 (AP 기준) ===")
print(results_df.round(4).to_string(index=False))

print(f"\nbest_params_: {gs.best_params_}")
print(f"best_score_ (AP): {gs.best_score_:.4f}")

# 4) 선택된 모델의 F1도 같은 5겹 분할에서 측정 (내부 추정치임을 명시)
f1_scores = cross_val_score(gs.best_estimator_, X_train, y_train, cv=cv, scoring="f1")
print(f"\n선택된 모델의 F1 (같은 5겹 CV, 내부 추정치): {f1_scores.mean():.4f}")
print("※ 이 값은 모델 선택에 이미 사용된 데이터에서 나온 추정치이므로,")
print("  최종 일반화 성능으로 보고하지 않고, 반드시 X_test로 별도 검증해야 함")


=== 조합별 CV 점수 (AP 기준) ===
 param_pre__cat__min_frequency  param_clf__min_samples_leaf  mean_test_score  std_test_score
                            10                            5           0.7582          0.0170
                            50                            5           0.7561          0.0169
                            50                           20           0.7535          0.0192
                            10                           20           0.7512          0.0185
                            10                           50           0.7404          0.0198
                            50                           50           0.7399          0.0195

best_params_: {'clf__min_samples_leaf': 5, 'pre__cat__min_frequency': 10}
best_score_ (AP): 0.7582

선택된 모델의 F1 (같은 5겹 CV, 내부 추정치): 0.6560
※ 이 값은 모델 선택에 이미 사용된 데이터에서 나온 추정치이므로,
  최종 일반화 성능으로 보고하지 않고, 반드시 X_test로 별도 검증해야 함


## 모델 카드 v5 — 구매 전환 예측

- 데이터: UCI Online Shoppers (12,330세션, 원본 예측 피처 17개 검토)
- 검증 방식: StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
- 주 지표: AP(`average_precision`) · F1 (정확도는 참고)
- **전처리: ColumnTransformer — 수치 {10}개 passthrough / 범주 {7}개 One-Hot(min_frequency={10})**
  - **변환 후 피처 {70}개**

- **파생변수(가설 → 결과)**
  - 가설 A `has_page_value`: { PageValues가 0인지 여부 자체가 관심 페이지 방문이라는 단순하고 강한 신호로 작동할 수 있다 } → Δf1 {+0.0744} · ΔAP {+0.0153} → **{ 채택}**
  - 가설 B `total_pages`: { 세 페이지 유형의 총 탐색량이 클수록 구매와 관련 있을 수 있다} → Δf1 {-0.0047} · ΔAP {+0.0017} → **{ 기각 }**

- **누수 점검: Month×TrafficType×Region 조합 키(715개, 65%가 10건 미만)를 전체 데이터로 미리 인코딩해 본 결과 F1이 +0.0373 부풀려짐 → TargetEncoder를 Pipeline 안에서 교차 적합한 버전은 기준(F1=0.5714) 대비 오히려 -0.0125 낮아 **이 피처는 최종 기각**
  - **학습되는 전처리의 `fit`이 `Pipeline` 안에서 일어나는가: { 예 }**

- **튜닝: GridSearchCV(scoring="average_precision"), 격자 {pre__cat__min_frequency: [10,50], clf__min_samples_leaf: [5,20,50]} → best {min_frequency=10, min_samples_leaf=5}**

- 내부 CV 추정: AP 0.7582 ± 0.0170 / F1 0.6560 ± 0.0230

- 최종 평가: { 아직 미실시 — X_test(전체의 20%, 튜닝에 전혀 사용 안 됨)로 단 한 번의 최종 검증이 필요함 }

- **산출물: purchase_pipeline.joblib (Pipeline 통째, compress=3) · experiment_log_v5.csv**
  ① `PageValues`가 실제 서비스에서 "구매 시점 이전에" 확정 가능한 값인지 확인 필요 
     (세션 종료 후에만 알 수 있는 값이라면 타깃 누수 위험) 
  ② 지금까지의 모든 검증은 내부 CV뿐이므로, 외부(다른 기간·다른 트래픽) 데이터로의 
     일반화 검증 필요 
  ③ RandomForest는 성능은 좋지만 해석이 어려우므로, 피처 중요도나 부분의존도 
     플롯으로 실무자에게 설명 가능한 형태로 정리 필요
- AI 사용 내역: Claude에게 ColumnTransformer+Pipeline 설계, 파생변수 가설 검증 코드, 
  타깃 인코딩 누수 실험, GridSearchCV 격자 탐색 코드를 요청함. 각 단계마다 
  `named_steps`/`cv_results_`로 실제 fit 위치와 조합별 점수를 직접 확인하며 
  Pipeline 내부에서 전처리가 fold별로 올바르게 재학습되는지 검증함.